internal/sbi/httpcallback/api_handle_dereg_notification.go

```diff
@@ -34,105 +33,118 @@ func HTTPAmfHandleDeregistrationNotification(c *gin.Context) {
 
 	err = openapi.Deserialize(&deregData, requestBody, "application/json")
 	if err != nil {
-		problemDetails := "[Request Body] " + err.Error()
-		rsp := models.ProblemDetails{
+		problemDetails := models.ProblemDetails{
 			Title:  "Malformed request syntax",
 			Status: http.StatusBadRequest,
-			Detail: problemDetails,
+			Detail: "[Request Body] " + err.Error(),
 		}
-		logger.CallbackLog.Errorln(problemDetails)
-		c.JSON(http.StatusBadRequest, rsp)
+		logger.CallbackLog.Errorln(problemDetails.Detail)
+		c.JSON(http.StatusBadRequest, problemDetails)
 		return
 	}
 
 	ueid := c.Param("ueid")
-	amfSelf := amf_context.GetSelf()
-	ue, ok := amfSelf.AmfUeFindByUeContextID(ueid)
+	deregType := c.Query("type")
+	ue, ok := amf_context.GetSelf().AmfUeFindByUeContextID(ueid)
 	if !ok {
 		logger.CallbackLog.Errorf("AmfUe Context[%s] not found", ueid)
-		problemDetails := &models.ProblemDetails{
+		problemDetails := models.ProblemDetails{
 			Status: http.StatusNotFound,
 			Cause:  "CONTEXT_NOT_FOUND",
 		}
 		c.JSON(http.StatusNotFound, problemDetails)
 		return
 	}
 
-	switch deregData.DeregReason {
-	case models.DeregistrationReason_UE_INITIAL_REGISTRATION:
-		// TS 23.502 - 4.2.2.2.2 General Registration
-		// Invokes the Nsmf_PDUSession_ReleaseSMContext for the corresponding access type
-		ue.SmContextList.Range(func(key, value interface{}) bool {
-			smContext := value.(*amf_context.SmContext)
+	problemDetails, err := DeregistrationNotificationProcedure(ue, deregType, deregData)
+	if problemDetails != nil {
+		ue.GmmLog.Errorf("Deregistration Notification Procedure Failed Problem[%+v]", problemDetails)
+	} else if err != nil {
+		ue.GmmLog.Errorf("Deregistration Notification Procedure Error[%v]", err.Error())
+	}
+	// TS 23.503 - 5.3.2.3.2 UDM initiated NF Deregistration
+	// The AMF acknowledges the Nudm_UECM_DeRegistrationNotification to the UDM.
+	c.JSON(http.StatusNoContent, nil)
+}
+
+// TS 23.502 - 4.2.2.3.3 Network-initiated Deregistration
+// The AMF can initiate this procedure for either explicit (e.g. by O&M intervention) or
+// implicit (e.g. expiring of Implicit Deregistration timer)
+func DeregistrationNotificationProcedure(ue *amf_context.AmfUe, deregType string, deregData models.DeregistrationData) (
+	problemDetails *models.ProblemDetails, err error,
+) {
+	var doUecmDereg bool = true
 
-			if smContext.AccessType() == deregData.AccessType {
-				var problemDetails *models.ProblemDetails
-				problemDetails, err = consumer.SendReleaseSmContextRequest(ue, smContext, nil, "", nil)
-				if problemDetails != nil {
-					ue.GmmLog.Errorf("Release SmContext Failed Problem[%+v]", problemDetails)
-				} else if err != nil {
-					ue.GmmLog.Errorf("Release SmContext Error[%v]", err.Error())
+	if deregType == "implicit" {
```

Comment:
The implicit and explicit dereg procedures are almost same, only difference is nas msg sent in explicit dereg. The procedures before `ue.Remove()` can move out this `if`.

Reply:
> The implicit and explicit dereg procedures are almost same, only difference is nas msg sent in explicit dereg. The procedures before ue.Remove() can move out this if.

OK